sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -1225,62 +1225,174 @@ private async Task RunProcessingAsync(CancellationToken cancellationToken)
-                                                        CancellationToken cancellationToken)
+        /// <returns><c>true</c> if the <paramref name="partitionId"/> was owned and was being processed; otherwise, <c>false</c>.</returns>
+        ///
+        /// <remarks>
+        ///   Exceptions encountered when stopping processing for an owned partition will be logged and will result in the error handler
+        ///   being invoked.  They will not be surfaced to callers.  This is intended to be a safe operation consumed
+        ///   as part of the load balancing cycle, which is failure-tolerant.
+        /// </remarks>
+        ///
+        private async Task<bool> TryStopProcessingPartitionAsync(string partitionId,
+                                                                 ProcessingStoppedReason reason,
+                                                                 CancellationToken cancellationToken)
         {
-            // TODO:  Implement this;  current code is just a place holder.
-            if (ActivePartitionProcessors.TryRemove(partitionId, out var processor))
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            Logger.EventProcessorPartitionProcessingStop(partitionId, Identifier, EventHubName, ConsumerGroup);
+
+            var partition = default(TPartition);
+
+            try
             {
-                processor.CancellationSource.Cancel();
+                // If the partition processor is not being tracked or could not be retrieved from the tracking items,
+                // then it cannot be stopped.
+
+                if (!ActivePartitionProcessors.TryRemove(partitionId, out var partitionProcessor))
+                {
+                    return false;
+                }
+
+                // Attempt to stop the processor; any exceptions should be treated as a problem with processing, not
+                // associated with the attempt to stop.
+
+                partition = partitionProcessor.Partition;
 
                 try
                 {
-                    await processor.ProcessingTask.ConfigureAwait(false);
+                    partitionProcessor.CancellationSource.Cancel();
+                    await partitionProcessor.ProcessingTask.ConfigureAwait(false);
                 }
                 catch (TaskCanceledException)
                 {
-                    // nop
+                    // This is expected; no action is needed.
                 }
-                finally
+                catch
                 {
-                    processor.CancellationSource.Dispose();
+                    // The processing task is in a failed state; any logging and dispatching
+                    // to the error handler happened in the processing task before the exception
+                    // was thrown.  All that remains is to override the reason for stopping.
+
+                    reason = ProcessingStoppedReason.OwnershipLost;
```

Comment:
Can you explain to me why we change the reason here?  I don't understand why the task being in a faulted state by the time we observe it means we change the reason to OwnershipLost.  It feels like this is subtle - so I'm just trying to understand the mental model.

Reply:
I'll have to defer to @kinelski on this; it was part of the cross-language design for the processor for which I was only on the fringe.  I remember pushing for an `Error` or similar reason but the team deciding against it.   I don't recall the reasoning, however.